base copy for Unreal + adding references everywhere (#312)
configcat/docs: website/docs/sdk-reference/unreal.md

---
id: unreal
title: Unreal Engine SDK Reference
description: ConfigCat Unreal Engine SDK Reference. This is a step-by-step guide on how to use feature flags in your Unreal Engine project.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
export const CPPSchema = require('@site/src/schema-markup/sdk-reference/cpp.json');

<script
  type="application/ld+json"
  dangerouslySetInnerHTML={{ __html: JSON.stringify(CPPSchema) }}
></script>

[![Star on GitHub](https://img.shields.io/github/stars/configcat/unreal-engine-sdk.svg?style=social)](https://github.com/configcat/unreal-engine-sdk/stargazers)
[![Build Status](https://img.shields.io/github/actions/workflow/status/configcat/unreal-engine-sdk/plugin-ci.yml?logo=GitHub&label=Unreal&branch=main)](https://github.com/configcat/unreal-engine-sdk/actions/workflows/plugin-ci.yml)

<a href="https://github.com/ConfigCat/unreal-engine-sdk" target="_blank">ConfigCat Unreal SDK on GitHub</a>

## Getting Started:

### 1. Installing the ConfigCat plugin to your project

Via **[Unreal Marketplace](https://www.unrealengine.com/marketplace/product/e142293a397d4ce4bf6a1f3053a2316d)**

> Via **[Github clone](https://github.com/configcat/unreal-engine-sdk)**

Reviewer: ```suggestion
Via **[GitHub clone](https://github.com/configcat/unreal-engine-sdk)**
```

| ----------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `Sdk Key`                     | SDK Key to access your feature flags and configurations. Get it from ConfigCat Dashboard.                                                                                                                                                                                                           |
| `Base Url`                    | Optional, sets the CDN base url (forward proxy, dedicated subscription) from where the sdk will download the config JSON.                                                                                                                                                                        |

> | `Data Governance`             | Optional, defaults to `Global`. Describes the location of your feature flag and setting data within the ConfigCat CDN. This parameter needs to be in sync with your Data Governance preferences. [More about Data Governance](advanced/data-governance.md). Available options: `Global`, `EuOnly`.  |

Reviewer 2: ```suggestion
| `Data Governance`             | Optional, defaults to `Global`. Describes the location of your feature flag and setting data within the ConfigCat CDN. This parameter needs to be in sync with your Data Governance preferences. [More about Data Governance](/advanced/data-governance). Available options: `Global`, `EuOnly`.  |
```
The internal link format changed a bit, and the correct ones should start with a / and not end with .md

| `ConnectTimeoutMs`      | Optional, defaults to `8000ms`. Sets the amount of milliseconds to wait for the server to make the initial connection (i.e. completing the TCP connection handshake). `0` means it never times out during transfer                                                                                  |
| `ReadTimeoutMs`         | Optional, defaults to `5000ms`. Sets the amount of milliseconds to wait for the server to respond before giving up. `0` means it never times out during transfer.                                                                                                                                   |
| `PollingMode`           | Optional, sets the polling mode for the client. [More about polling modes](#polling-modes).                                                                                                                                                                                                         |

> | `AutoPollInterval`      | For PollingMode == Custom, sets at least how often this policy should fetch the latest configuration and refresh the cache.                                                                                                                                                                         |

Reviewer 3: ```suggestion
| `AutoPollInterval`      | For PollingMode == Custom, sets at least how often this policy should fetch the latest config JSON and refresh the cache.                                                                                                                                                                         |
```

Author: I missed this suggestion, I will apply the change manually next commit.

> | `Auto Poll Interval`          | For PollingMode == Custom, sets at least how often this policy should fetch the latest config JSO and refresh the cache.                                                                                                                                                                            |

Reviewer 3: ```suggestion
| `Auto Poll Interval`          | For PollingMode == Custom, sets at least how often this policy should fetch the latest config JSON and refresh the cache.                                                                                                                                                                           |
```

FString MyFourthFeatureFlag = ConfigCat->GetStringValue(TEXT("myForthFeatureFlag"), TEXT(""));
```

Access value depending on type targeting an user:
```cpp
UConfigCatSubsystem* ConfigCat = UConfigCatSubsystem::Get(this);

FConfigCatUser User = FConfigCatUser(TEXT("#USER-IDENTIFIER#"));
FString TargetValue = ConfigCat->GetStringValue(TEXT("targetValue"), TEXT(""), User);
```

</TabItem>
</Tabs>

## Anatomy of `GetValueDetails()`

`GetValueDetails()` is similar to `GetValue()` but instead of returning the evaluated value only, it gives more detailed information about the evaluation result.

| Parameters      | Description                                                                                                  |
| --------------- | ------------------------------------------------------------------------------------------------------------ |
| `Key`           | **REQUIRED.** The key of a specific setting or feature flag. Set on _ConfigCat Dashboard_ for each setting.  |
| `Default Value` | **REQUIRED.** This value will be returned in case of an error.                                               |

> | `User`          | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

Reviewer 2: ```suggestion
| `User`          | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](/advanced/targeting) |
```

| `Value`                              | The evaluated value of the feature flag or setting.                                       |
| `Key`                                | The key of the evaluated feature flag or setting.                                         |
| `Is Default Value`                   | True when the default value passed to getValueDetails() is returned due to an error.      |
| `Error`                              | In case of an error, this field contains the error message.                               |
| `User`                               | The user object that was used for evaluation.                                             |
| `Matched Evaluation Percentage Rule` | If the evaluation was based on a percentage rule, this field contains that specific rule. |
| `Matched Evaluation Rule`            | If the evaluation was based on a targeting rule, this field contains that specific rule.  |
| `FetchTime`                          | The last download time of the current config.                                             |

## User Object

> The [User Object](../advanced/user-object.md) is essential if you'd like to use ConfigCat's [Targeting](advanced/targeting.md) feature.

Reviewer 2: ```suggestion
The [User Object](/advanced/user-object) is essential if you'd like to use ConfigCat's [Targeting](/advanced/targeting) feature.
```

<img className="unreal-settings-overrides zoomable" src="/docs/assets/unreal/settings-overrides.png" alt="Unreal Engine Settings Overrides" />

### JSON File

The SDK can be set up to load your feature flag & setting overrides from a file.

#### File

If specified the SDK will load the JSON data `%PROJECT_ROOT/Content/ConfigCat/flags.json` file. This file needs to be created manually in the specified folder to ensure it gets packaged in the final executable. 

Note: This file needs to be created in your file explorer and cannot be done within Unreal.

#### JSON File Structure

The SDK supports 2 types of JSON structures to describe feature flags & settings.

##### 1. Simple (key-value) structure

```json
{
  "flags": {
    "enabledFeature": true,
    "disabledFeature": false,
    "intSetting": 5,
    "doubleSetting": 3.14,
    "stringSetting": "test"
  }
}
```

##### 2. Complex (full-featured) structure

This is the same format that the SDK downloads from the ConfigCat CDN.
It allows the usage of all features you can do on the ConfigCat Dashboard.

You can download your current config JSON from ConfigCat's CDN and use it as a baseline.

> The URL to your current config JSON is based on your [Data Governance](advanced/data-governance.md) settings:

Reviewer 2: ```suggestion
The URL to your current config JSON is based on your [Data Governance](/advanced/data-governance) settings:
```

</TabItem>
<TabItem value="cpp" label="C++">

```cpp
UConfigCatSubsystem* ConfigCat = UConfigCatSubsystem::Get(this);
TMap<FString, FConfigCatValue> SettingValues = ConfigCat->GetAllValues(); 
	
FConfigCatUser User = FConfigCatUser(TEXT("#UNIQUE-USER-IDENTIFIER#"));
TMap<FString, FConfigCatValue> SettingValuesTargeting = ConfigCat->GetAllValues(User);
```

</TabItem>
</Tabs>

## `GetAllValueDetails`

Evaluates and returns the detailed values of all feature flags and settings. Passing a [User Object](#user-object) is optional.

<Tabs groupId="unreal-languages">
<TabItem value="blueprints" label="Blueprints">

<img className="unreal-blueprints-get-all-value-details zoomable" src="/docs/assets/unreal/blueprints-get-all-value-details.png" alt="Unreal Engine Get All Value Details" />

</TabItem>
<TabItem value="cpp" label="C++">

```cpp
UConfigCatSubsystem* ConfigCat = UConfigCatSubsystem::Get(this);
FConfigCatUser User = FConfigCatUser(TEXT("#UNIQUE-USER-IDENTIFIER#"));

TArray<FConfigCatEvaluationDetails> AllValueDetails = ConfigCat->GetAllValueDetails(User);
```

</TabItem>
</Tabs>

## Custom Cache

> **Coming soon**: [https://github.com/configcat/unreal-engine-sdk/issues/9](Support for Custom Cache #9)

Reviewer: <img width="912" alt="image" src="https://github.com/configcat/docs/assets/58779216/8366afe3-1e59-480f-b469-60c69c15728e">

Improperly formatted link

<TabItem value="blueprints" label="Blueprints">

<img className="unreal-blueprints-force-refresh zoomable" src="/docs/assets/unreal/blueprints-force-refresh.png" alt="Unreal Engine Blueprint Force Refresh" />

</TabItem>
<TabItem value="cpp" label="C++">

```cpp
UConfigCatSubsystem* ConfigCat = UConfigCatSubsystem::Get(this);

ConfigCat->ForceRefresh();
```

</TabItem>
</Tabs>

## Using ConfigCat behind a proxy

:::info
The Unreal Engine HTTP module doesn't support running behinde a proxy.
:::

## Changing the default HTTP timeout

Set the maximum wait time for a ConfigCat HTTP response by changing the _ConnectTimeoutMs_ or _ReadTimeoutMs_ in the ConfigCat `ProjectSettings`.
The default _ConnectTimeoutMs_ is 8 seconds. The default _ReadTimeoutMs_ is 5 seconds.

<img className="unreal-settings-http-timeout zoomable" src="/docs/assets/unreal/settings-http-timeout.png" alt="Unreal Engine Settings Http Timeout" />

## Logging

All ConfigCat logs are inside the `LogConfigCat` category. By default the verbosity is set to `Log`, but it can be changed to any [Verbosity Level](https://docs.unrealengine.com/5.3/en-US/logging-in-unreal-engine/#logverbosity).

The verbosity can be changed:

> Via [Command line arguments]:

Reviewer: ```suggestion
Via **Command line arguments**:
```

Run the executable with `-LogCmds="LogConfigCat VerbosityLevel"`. For example: `-LogCmds="LogConfigCat Warning"` to only show warnings and above.

> Via [DefaultEngine.ini]:

Reviewer: ```suggestion
Via **DefaultEngine.ini**:
```

In the `[Core.Log]` category add `LogConfigCat=VerbosityLevel` inside the `DefaultEngine.ini`. For example:

```
[Core.Log]
LogConfigCat=Warning ;to only show warnings and above
```

Note: Since the Unreal Engine SDK is a wrapper of the CPP SDK, all logs coming from the CPP SDK are tagged wtih `[CPP-SDK]`.

Info level logging helps to inspect how a feature flag was evaluated:

```bash
[Info]: Evaluating getValue(isPOCFeatureEnabled)
User object: {
    "Email": "john@example.com",
    "Identifier": "435170f4-8a8b-4b67-a723-505ac7cdea92",
}
Evaluating rule: [Email:john@example.com] [CONTAINS] [@something.com] => no match
Evaluating rule: [Email:john@example.com] [CONTAINS] [@example.com] => match, returning: true
```

## Sensitive information handling

> The frontend/mobile SDKs are running in your users' browsers/devices. The SDK is downloading a [config JSON](/requests/) file from ConfigCat's CDN servers. The URL path for this config JSON file contains your SDK key, so the SDK key and the content of your config JSON file (feature flag keys, feature flag values, targeting rules, % rules) can be visible to your users.

Reviewer: ```suggestion
The frontend/mobile SDKs are running in your users' browsers/devices. The SDK is [downloading a config JSON](/requests/) file from ConfigCat's CDN servers. The URL path for this config JSON file contains your SDK key, so the SDK key and the content of your config JSON file (feature flag keys, feature flag values, targeting rules, % rules) can be visible to your users.
```

<img className="unreal-plugin-settings zoomable" src="/docs/assets/unreal/plugin-settings.png" alt="Unreal Engine plugin settings" />

| Properties              | Description                                                                                                                                                                                                                                                                                         |
| ----------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `SdkKey`                | SDK Key to access your feature flags and configurations. Get it from ConfigCat Dashboard.                                                                                                                                                                                                           |

> | `BaseUrl`               | Optional, sets the CDN base url (forward proxy, dedicated subscription) from where the sdk will download the configurations.                                                                                                                                                                        |

Reviewer 3: ```suggestion
| `BaseUrl`               | Optional, sets the CDN base url (forward proxy, dedicated subscription) from where the sdk will download the config JSON.                                                                                                                                                                        |
```
